packages/material-ui/src/Avatar/Avatar.d.ts

```diff
@@ -14,6 +14,6 @@ export interface AvatarProps
 
 export type AvatarClassKey = 'root' | 'colorDefault' | 'img';
 
-declare const Avatar: React.ComponentType<AvatarProps>;
+declare class Avatar<C> extends React.Component<C & AvatarProps<C>> {}
 
 export default Avatar;
```

Comment:
It seems like what we actually want here (and similarly for other components) is something more like

```ts
export interface AvatarProps<C = React.HTMLAttributes<HTMLDivElement>>
    extends StandardProps<C, AvatarClassKey> & {
```

I.e. in each case take the type we used to be passing as the first argument to `StandardProps`, make that the default for `C`, and then pass `C` to `StandardProps` instead. There's a problem though: we can no longer use an interface because we're intersecting with a type variable, so we have to switch to a type alias and intersection:

```ts
export type AvatarProps<C = React.HTMLAttributes<HTMLDivElement>>
    = StandardProps<C, AvatarClassKey> & {
```

Also `StandardProps` does the work of `&`-ing that first argument with the rest of your props, so you don't need to do it in the `declare` statement:

```ts
declare class Avatar<C> extends React.Component<AvatarProps<C>> {}
```

This approach looks promising at first blush, but it doesn't quite work ideally because the default type argument isn't used when you use a bare component, i.e. this fails:

```ts
<Avatar title="hello" /> // `title` from HTMLAttributes is invalid
```

You have to provide an explicit type parameter to fix it:

```ts
<Avatar<React.HTMLAttributes<HTMLDivElement>> title="hello" alt="Image Alt" src="example.jpg" />
```

This would be a huge breaking change and is just gross 😕

But conditional types to the rescue! This seems to work:

```ts
export type AvatarProps<C = {}> = StandardProps<
  {} extends C ? React.HTMLAttributes<HTMLDivElement> : C,
  AvatarClassKey
> & {
```

Basically, if `C` is not inferred as anything specific, either because you referenced bare `AvatarProps` or because you didn't provide a `component` prop to `Avatar` to allow it to disambiguate, you will get the prop types that match the default component, in this case a `div`.

Now we can just write

```ts
<Avatar title="hi" />
```

and importantly, this _doesn't_ work, which it shouldn't!

```ts
<Avatar component={(props: {/* no title prop! */}) => <div />} title="hello" />
```

We should be able to bake this pattern into the `StandardProps` operator, so you can just write

```ts
export type AvatarProps<C = {}> = StandardProps<C, AvatarClassKey> & {
```

What do you think?